### ISSUE TITLE
autofs module forces use of puppetlabs/concat 2.x

### ISSUE THREAD

**Issue by jgreen210** [voxpupuli/puppet-autofs#48]
## Affected Puppet, Ruby, OS and module versions/distributions

- Puppet: 4.9.4
- Ruby: 2.3.1
- Distribution: ubuntu
- Module version: 3.0.1

## How to reproduce (e.g Puppet code you use)

```
    ::autofs::mount { "share":
        mount => "/share",
        mapcontents => ["files -fstype=nfs4,ro,hard,intr,nosuid,exec hostname:/share/files"],
        options => "--timeout=600",
        order => 50
      }
```

## What are you seeing

I'm seeing one of these puppet warning for every puppet resource I have, which is making the puppet apply logs unreadable.

```
Warning: This method is deprecated, please use match expressions with Stdlib::Compat::Bool instead. They are described at https://docs.puppet.com/puppet/latest/reference/lang_data_type.html#match-expressions. at ["/tmp/puppet-deploy/manifests/.modules-contrib/concat/manifests/init.pp", 82]:
(at /tmp/puppet-deploy/manifests/.modules-contrib/stdlib/lib/puppet/functions/deprecation.rb:25:in `deprecation')
```

## What behaviour did you expect instead

No warnings. As much flexibility as possible for puppetlabs/concat dependency version.

## Any additional information you'd like to impart

puppet/autofs declares a dependency on puppetlabs/concat 2.x:

https://github.com/voxpupuli/puppet-autofs/blob/85b0e8c367cc19c77bba8bcc63f1bc2797c3e497/metadata.json#L104

...which means I am using puppetlabs/concat 2.2.1 - the latest 2.x release. ~~It looks like the above warnings from puppetlabs/concat might be fixed by puppetlabs/concat 3.x or 4.x~~. Is is possible for puppet-autofs to allow use a wider range of puppetlabs/concat versions?

**Comment by jgreen210** [comment]
The warning is coming from here (in concat 2.2.1):

https://github.com/puppetlabs/puppetlabs-concat/blob/5c4a9141d08a7b23dcada029d23b82590632d0f4/manifests/init.pp#L82

**Comment by jgreen210** [comment]
I made this change in a local fork of puppet-autofs 3.0.1:

```
commit bf414f84b9bcfe6f111012f8445c4717b4870fd0
Author: Jeremy Green <jeremy.green@tomtom.com>
Date:   Tue May 9 14:34:00 2017 +0100

    Allow puppetlabs/concat 2.x, 3.x or 4.x. https://github.com/voxpupuli/puppet-autofs/issues/48

diff --git a/metadata.json b/metadata.json
index da1b61d..ad2fc09 100644
--- a/metadata.json
+++ b/metadata.json
@@ -101,7 +101,7 @@
     },
     {
       "name": "puppetlabs/concat",
-      "version_requirement": ">= 2.0.0 < 3.0.0"
+      "version_requirement": ">= 2.0.0 < 5.0.0"
     }
   ]
 }
```

...which means I now use concat 4.0.0 (the latest). The above warning from concat module has been replaced with a slightly different one:

```
Warning: This method is deprecated, please use match expressions with Stdlib::Compat::String instead. They are described at https://docs.puppet.com/puppet/latest/reference/lang_data_type.html#match-expressions. at ["/tmp/puppet-deploy/manifests/.modules-contrib/concat/manifests/fragment.pp", 18]:
(at /tmp/puppet-deploy/manifests/.modules-contrib/stdlib/lib/puppet/functions/deprecation.rb:25:in `deprecation')
```

**Comment by jgreen210** [comment]
So:

* puppet/autofs should allow a wider range of puppetlabs/concat versions (assuming tests pass etc.)
* It's not clear to me yet what's making puppetlabs/concat emit warnings - autofs (which does use it), some other puppet module, or my own manifests.

**Comment by jgreen210** [comment]
The error I mention above:

```
Warning: This method is deprecated, please use match expressions with Stdlib::Compat::String instead. They are described at https://docs.puppet.com/puppet/latest/reference/lang_data_type.html#match-expressions. at ["/tmp/puppet-deploy/manifests/.modules-contrib/concat/manifests/fragment.pp", 18]:
(at /tmp/puppet-deploy/manifests/.modules-contrib/stdlib/lib/puppet/functions/deprecation.rb:25:in `deprecation')
```

...from autofs 3.0.1, patched to allow use of puppetlabs/concat 4.0.0, and using puppetlabs/concat 4.0.0 is definitely caused by autofs, not some other module/resource - I commented all other resources out, and was able to make the warning come and go just by commenting out the ::autofs::mount resource I mention above.

I can create a separate ticket for this if you want.

**Comment by dhollinger** [comment]
I'm actually preparing a PR for that now
